okio/src/main/kotlin/okio/Utf8.kt

```diff
@@ -110,89 +110,354 @@ fun String.utf8Size(beginIndex: Int = 0, endIndex: Int = length): Long {
+          yield((codePoint shr 18          or 0xf0).toByte()) // 11110xxx
+          yield((codePoint shr 12 and 0x3f or 0x80).toByte()) // 10xxxxxx
+          yield((codePoint shr 6  and 0x3f or 0x80).toByte()) // 10xxyyyy
+          yield((codePoint        and 0x3f or 0x80).toByte()) // 10yyyyyy
+          /* ktlint-enable no-multi-spaces */
+          index += 2
+        }
+      }
+    }
+  }
 }
 
-internal fun isIsoControl(codePoint: Int): Boolean =
-  (codePoint in 0x00..0x1F) || (codePoint in 0x7F..0x9F)
+// TODO combine with Buffer.readUtf8CodePoint?
+internal inline fun ByteArray.processUtf8CodePoints(
+  beginIndex: Int,
+  endIndex: Int,
+  yield: (Int) -> Unit
+) {
+  var index = beginIndex
+  while (index < endIndex) {
+    val b0 = this[index]
+    when {
+      b0 >= 0 -> {
+        // 0b0xxxxxxx
+        yield(b0.toInt())
+        index++
+
+        // Assume there is going to be more ASCII
+        while (index < endIndex && this[index] >= 0) {
+          yield(this[index++].toInt())
+        }
+      }
+      b0 shr 5 == -2 -> {
+        // 0b110xxxxx
+        index += process2Utf8Bytes(index, endIndex) { yield(it) }
+      }
+      b0 shr 4 == -2 -> {
+        // 0b1110xxxx
+        index += process3Utf8Bytes(index, endIndex) { yield(it) }
+      }
+      b0 shr 3 == -2 -> {
+        // 0b11110xxx
+        index += process4Utf8Bytes(index, endIndex) { yield(it) }
+      }
+      else -> {
+        // 0b10xxxxxx - Unexpected continuation
+        // 0b111111xxx - Unknown encoding
+        yield(REPLACEMENT_CODE_POINT)
+        index++
+      }
+    }
+  }
+}
+
+// TODO combine with Buffer.readUtf8?
+internal inline fun ByteArray.processUtf8Chars(
+  beginIndex: Int,
+  endIndex: Int,
+  yield: (Char) -> Unit
+) {
+  var index = beginIndex
+  while (index < endIndex) {
+    val b0 = this[index]
+    when {
+      b0 >= 0 -> {
+        // 0b0xxxxxxx
+        yield(b0.toChar())
+        index++
+
+        // Assume there is going to be more ASCII
+        // This is almost double the performance of the outer loop
+        while (index < endIndex && this[index] >= 0) {
+          yield(this[index++].toChar())
+        }
+      }
+      b0 shr 5 == -2 -> {
+        // 0b110xxxxx
+        index += process2Utf8Bytes(index, endIndex) { yield(it.toChar()) }
+      }
+      b0 shr 4 == -2 -> {
+        // 0b1110xxxx
+        index += process3Utf8Bytes(index, endIndex) { yield(it.toChar()) }
+      }
+      b0 shr 3 == -2 -> {
+        // 0b11110xxx
+        index += process4Utf8Bytes(index, endIndex) { codePoint ->
+          if (codePoint != REPLACEMENT_CODE_POINT) {
+            // Unicode code point:    00010000000000000000 + xxxxxxxxxxyyyyyyyyyy (21 bits)
+            // UTF-16 high surrogate: 110110xxxxxxxxxx (10 bits)
+            // UTF-16 low surrogate:  110111yyyyyyyyyy (10 bits)
+            /* ktlint-disable no-multi-spaces */
+            yield(((codePoint ushr 10   ) + (0xd800 - (0x010000 ushr 10))).toChar())
```

Comment:
I wonder how much of this gets evaluated at build time

Reply:
I'm not sure. I bet the Java JIT compiler quickly does, but I'm not sure how it would perform on JS or Native. I could pull it out as a constant to make sure.